### PR TITLE
Extract quote reification from Staging

### PR DESCRIPTION
**Pull request by nicolasstucki**
Split `Staging` into
- [x] `Staging`: Check PCP, heal types and expand macros
- [x] `ReifyQuotes`: Transform top level quotes and pickles their contents
- [ ] Move `Staging` before `Pickler`

### CODE REVIEW

**Comment by nicolasstucki** [comment]
Test performance please

**Comment by nicolasstucki** [comment]
test performance please

File: compiler/src/dotty/tools/dotc/transform/ReifyQuotes.scala [hunk 1]
```diff
@@ -0,0 +1,425 @@
...
+        assert(!tree.symbol.isQuote)
+        // assert(!tree.symbol.isSplice) // TODO widen ~ type references at stage 0?
+        assert(tree.symbol != defn.QuotedExpr_~)
+      case tree: Select if tree.symbol == defn.QuotedExpr_~ =>
+        assert(Splicer.canBeSpliced(tree.qualifier))
+      case _ =>
+    }
+  }
+
+  override def run(implicit ctx: Context): Unit =
```
Comment (allanrenucci) [inline compiler/src/dotty/tools/dotc/transform/ReifyQuotes.scala]: I think you should override `isRunnable` and not `run`:
```scala
override def isRunnable(implicit ctx: Context): Boolean =
  super.isRunnable && ctx.compilationUnit.needsStaging
```

Comment (allanrenucci) [reply compiler/src/dotty/tools/dotc/transform/ReifyQuotes.scala]: Actually I am wrong. `isRunnable` is for the whole phase and not a single compilation unit

**Comment by biboudis** [comment]
Do you think we should close this?

**Comment by nicolasstucki** [comment]
Replaced with #5763

**Pull request closed without merging**